File: src-tauri/src/models/timeline.rs
use serde::{Deserialize, Serialize};

/// Serde default helper for `include_scenes` — `true` so omitting the field
/// keeps Scenes in the timeline (the common case). See ADR-0033.
fn default_true() -> bool {
    true
}

/// Input filter for the `query_timeline` aggregation command (ADR-0033).
///
/// All fields optional; the command is a read-only projection over the
/// existing `events` + `scenes` tables — there is NO dedicated timeline table
/// (ADR-0033: derived, never persisted, never authored).
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineQueryInput {
    /// Filter to entries involving this character (across any phase).
    #[serde(default)]
    pub character_id: Option<String>,

    /// Filter to entries at this location.
    #[serde(default)]
    pub location_id: Option<String>,

    /// Inclusive lower bound on `start_at` (ISO 8601 string compare).
    #[serde(default)]
    pub from: Option<String>,

    /// Exclusive upper bound on `start_at`.
    #[serde(default)]
    pub to: Option<String>,

    /// Restrict SCENES to this novel (does NOT affect events).
    #[serde(default)]
    pub novel_id: Option<String>,

    /// Restrict SCENES to those referencing this item (does not affect
    /// events — events have no items).
    #[serde(default)]
    pub item_id: Option<String>,

    /// Default `true`; `false` = events only.
    #[serde(default = "default_true")]
    pub include_scenes: bool,

    /// Default 50, clamped to [1, 100] by the command.
    #[serde(default)]
    pub limit: Option<i64>,
}

/// One chronological entry in the timeline — either an Event or a Scene
/// projected onto a common shape (ADR-0033).
///
/// The agent `timeline_lookup` surface returns every entry at its own
/// `start_at`; the UI's visual "absorption" rule is intentionally NOT applied
/// here (ADR-0033: chronological truth over visual de-duplication — this
/// divergence is the single most surprising property of the design).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineEntry {
    /// `"event"` or `"scene"`.
    pub kind: String,
    pub id: String,
    pub name: String,
    pub start_at: Option<String>,
    pub end_at: Option<String>,
    /// Resolved location name; `None` if the entry has no location.
    pub location_name: Option<String>,
    /// Character NAMES involved (resolved via the `*_character_refs` junction
    /// → `characters.name`). Empty `Vec` when none — never `None`.
    pub participants: Vec<String>,
    /// First ~200 chars of the event `description` / scene `summary`; `None`
    /// if empty.
    pub description_excerpt: Option<String>,
    /// EVENTS ONLY: names of scenes referencing this event via
    /// `scene_event_refs`. `None` for scenes.
    pub narrated_by_scene_names: Option<Vec<String>>,
    /// SCENES ONLY: names of events referenced by this scene via
    /// `scene_event_refs`. `None` for events.
    pub narrated_event_names: Option<Vec<String>>,
    /// SCENES ONLY: the parent novel's title. `None` for events.
    pub novel_title: Option<String>,
}

/// Response from `query_timeline`.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineResponse {
    /// Sorted ascending by `start_at`; entries with `NULL` `start_at` go LAST.
    pub entries: Vec<TimelineEntry>,
    /// Full match count BEFORE `limit` applied.
    pub total: i64,
    /// `total > entries.len()`.
    pub truncated: bool,
}

/// One row from `list_timeline_lanes` — a Character plus the number of
/// DISTINCT timeline entries (Events via `event_character_refs` + Scenes via
/// `scene_character_refs`) they participate in.
///
/// Drives the Timeline UI's default lane selection: lanes where
/// `participation_count > 2` are auto-shown, the rest populate the character
/// multiselect. Characters with zero participation are excluded by the query
/// (INNER JOIN) — they'd never be default-selected and would only clutter the
/// multiselect.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineLane {
    pub character_id: String,
    pub name: String,
    pub participation_count: i64,
}

File: src/types/timeline.ts
import { z } from "zod";

/**
 * Timeline — derived, read-only chronological projection of a World's Events
 * and Scenes (ADR-0033).
 *
 * The Timeline is never persisted and never authored: it is computed at query
 * time by the `query_timeline` Rust aggregation command, which cross-entity-
 * joins Events + Scenes with their resolved location names, participant names,
 * and cross-reference annotations into a single flat list.
 *
 * Unlike `list_events` (which returns Event rows), the Timeline returns
 * time-ordered, cross-entity-joined entries. Every Event and Scene appears at
 * its own `startAt` — events referenced by scenes are NOT deduplicated, so the
 * chronology is chronologically truthful (the tool surface intentionally
 * diverges from the UI surface's visual absorption rule).
 */

// ─── Query ────────────────────────────────────────────────────────────────

/**
 * Hard ceiling on `limit` — mirrored from `MAX_LIMIT` in
 * `src-tauri/src/commands/timeline.rs`; keep both in sync. The Timeline UI
 * steps toward this via its "load more" control.
 */
export const TIMELINE_LIMIT_MAX = 500;

/**
 * Filter parameters for a Timeline query. All fields optional; an empty query
 * returns the whole chronology (up to `limit`, default 50).
 */
export const timelineQuerySchema = z.object({
  /** Restrict to entries involving this character (by id). */
  characterId: z.string().optional(),
  /** Restrict to entries taking place at this location (by id). */
  locationId: z.string().optional(),
  /** Lower bound (inclusive) on `startAt`, as an ISO 8601 timestamp. */
  from: z.string().optional(),
  /** Upper bound (exclusive) on `startAt`, as an ISO 8601 timestamp. */
  to: z.string().optional(),
  /** Restrict scene entries to this novel (by id). Events are unaffected. */
  novelId: z.string().optional(),
  /** Restrict scene entries to those referencing this item (by id). Does not affect events — events have no items. */
  itemId: z.string().optional(),
  /** Whether to include Scene entries alongside Events (default `true`). */
  includeScenes: z.boolean().optional(),
  /** Maximum entries to return (default 50, max `TIMELINE_LIMIT_MAX`). */
  limit: z.number().int().min(1).max(TIMELINE_LIMIT_MAX).optional(),
});

export type TimelineQuery = z.infer<typeof timelineQuerySchema>;

// ─── Entry ────────────────────────────────────────────────────────────────

/**
 * A single node in the chronology — either an Event or a Scene, flattened into
 * a uniform shape with resolved (denormalized) names.
 *
 * The `id` is polymorphic: it is an Event id when `kind === "event"` and a
 * Scene id when `kind === "scene"`. It is therefore a plain `string`, not a
 * branded entity id.
 *
 * Cross-reference fields are optional and kind-specific:
 * - Events carry `narratedBySceneNames` (scenes that reference this event).
 * - Scenes carry `narratedEventNames` (events this scene references),
 *   `novelTitle` (the containing novel's title), and `novelId` / `chapterId`
 *   (the ids needed to navigate from a Scene node to its Novel chapter
 *   workspace — ADR-0034 additive extension).
 */
export const timelineEntrySchema = z.object({
  /** Discriminator: `"event"` or `"scene"`. */
  kind: z.enum(["event", "scene"]),
  /** Event or Scene UUID (polymorphic — not a branded entity id). */
  id: z.string(),
  /** Display name (Event name or Scene title). */
  name: z.string(),
  /** Story timeline — when the entry starts (ISO 8601). `null` if undated. */
  startAt: z.iso.datetime().nullable(),
  /** Story timeline — when the entry ends (ISO 8601). `null` if undated. */
  endAt: z.iso.datetime().nullable(),
  /** Resolved location name. `null` if no location is set. */
  locationName: z.string().nullable(),
  /** Resolved character names participating in this entry. */
  participants: z.array(z.string()),
  /** First ~200 characters of the Event description or Scene summary. */
  descriptionExcerpt: z.string().nullable(),
  /**
   * Names of Scenes that narrate / reference this Event. An array (possibly
   * empty) for events; `null` for scenes. Always present in the payload —
   * Rust serde emits the key with `null` for the non-applicable kind.
   */
  narratedBySceneNames: z.array(z.string()).nullable(),
  /** Names of Events this Scene references. An array for scenes; `null` for events. */
  narratedEventNames: z.array(z.string()).nullable(),
  /** Title of the Novel containing this Scene. A string for scenes; `null` for events. */
  novelTitle: z.string().nullable(),
  /** Id of the Novel containing this Scene. A string for scenes; `null` for events (ADR-0034). */
  novelId: z.string().nullable(),
  /** Id of the Chapter containing this Scene. A string for scenes; `null` for events (ADR-0034). */
  chapterId: z.string().nullable(),
});

export type TimelineEntry = z.infer<typeof timelineEntrySchema>;

// ─── Response ─────────────────────────────────────────────────────────────

/**
 * Result of a Timeline query. Entries are sorted ascending by `startAt`
 * (undated entries — `null` startAt — sort last). `total` reports the full
 * match count (ignoring `limit`); `truncated` is `true` when `entries.length`
 * was capped by `limit`.
 */
export const timelineResponseSchema = z.object({
  entries: z.array(timelineEntrySchema),
  /** Full match count (may exceed `entries.length` when truncated). */
  total: z.number().int(),
  /** Whether `limit` cut the result short. */
  truncated: z.boolean(),
});

export type TimelineResponse = z.infer<typeof timelineResponseSchema>;

// ─── Lane ─────────────────────────────────────────────────────────────────

/**
 * A Character plus their Timeline participation count — the number of
 * DISTINCT Events (via `event_character_refs`) plus Scenes (via
 * `scene_character_refs`) they appear in. A character in one event via two
 * phases counts as 1.
 *
 * Drives the Timeline UI's default lane selection: lanes where
 * `participationCount > 2` are auto-shown; the rest populate the character
 * multiselect. Characters with zero participation are excluded by the
 * backend query (they'd never be default-selected).
 *
 * Returned by `list_timeline_lanes`.
 */
export const timelineLaneSchema = z.object({
  characterId: z.string(),
  name: z.string(),
  participationCount: z.number().int(),
});

export type TimelineLane = z.infer<typeof timelineLaneSchema>;
